src/objectprotocol/number.rs

```diff
@@ -0,0 +1,323 @@
+use std::cmp::Ordering;
+use std::fmt;
+
+use crate::conversion::ToPyObject;
+use crate::err::{self, PyErr, PyResult};
+use crate::ffi;
+use crate::objects::{PyObject, PyInt, PyLong, PyFloat};
+use crate::python::{Python, PythonObject, ToPythonPointer};
+
+
+use super::ObjectProtocol;
+
+/// Operations on numeric objects
+pub trait NumberProtocol: ObjectProtocol {
+    /// Perform addition (self + other)
+    ///
+    /// Invokes the `__add__` magic-method
+    #[inline]
+    fn add(&self, py: Python, other: impl ToPyObject) -> PyResult<PyObject> {
+        other.with_borrowed_ptr(py, |other| unsafe {
+            err::result_from_owned_ptr(py, ffi::PyNumber_Add(self.as_ptr(), other))
+        })
+    }
+    /// Perform subtraction (self - other)
+    ///
+    /// Invokes the `__sub__` magic-method
+    #[inline]
+    fn subtract(&self, py: Python, other: impl ToPyObject) -> PyResult<PyObject> {
+        other.with_borrowed_ptr(py, |other| unsafe {
+            err::result_from_owned_ptr(py, ffi::PyNumber_Subtract(self.as_ptr(), other))
+        })
+    }
+    /// Perform multiplication (self * other)
+    ///
+    /// Invokes the `__mul__` magic-method
+    #[inline]
+    fn multiply(&self, py: Python, other: impl ToPyObject) -> PyResult<PyObject> {
+        other.with_borrowed_ptr(py, |other| unsafe {
+            err::result_from_owned_ptr(py, ffi::PyNumber_Multiply(self.as_ptr(), other))
+        })
+    }
+    /// Perform matrix multiplication, equivalent to the Python expression `self @ other`
+    ///
+    /// Invokes the `__matmul__` magic-method
+    ///
+    /// This was added in Python 3.5, and will unconditionally
+    /// throw an exception on any version before that.
+    ///
+    /// See [PEP 0456](https://www.python.org/dev/peps/pep-0465/) for details.
+    #[inline]
+    fn matrix_multiply(&self, py: Python, other: impl ToPyObject) -> PyResult<PyObject> {
+        #[cfg(not(any(feature = "python3-4", feature = "python2-sys")))] {
+            other.with_borrowed_ptr(py, |other| unsafe {
+                err::result_from_owned_ptr(py, ffi::PyNumber_MatrixMultiply(self.as_ptr(), other))
+            })
+        }
+        #[cfg(any(feature = "python3-4", feature = "python2-sys"))] {
```

Comment:
This feature checks if the user explicitly requested python3-4 as a cargo feature.
It won't be set when python 3.4 is instead selected via `PYTHON_SYS_EXECUTABLE`.

Fixing this will require solving https://github.com/dgrunwald/rust-cpython/pull/269#discussion_r696136064 in some way.